bugsnag-plugin-android-anr/src/main/jni/anr_handler.c

```diff
@@ -5,118 +5,115 @@
   // https://developer.android.com/training/articles/perf-jni#faq:-why-didnt-findclass-find-my-class
   int result = (*env)->GetJavaVM(env, &bsg_jvm);
   if (result != 0) {
-    // Failed to fetch VM, cannot continue
+    BUGSNAG_LOG("Failed to fetch Java VM. ANR handler not installed.");
     return false;
   }
 
   jclass clz = (*env)->FindClass(env, "com/bugsnag/android/AnrPlugin");
   mthd_notify_anr_detected =
-      (*env)->GetMethodID(env, clz, "notifyAnrDetected", "()V");
+          (*env)->GetMethodID(env, clz, "notifyAnrDetected", "()V");
   return true;
 }
 
-void bsg_handle_sigquit(int signum, siginfo_t *info, void *user_context) {
+static void notify_anr_detected() {
   if (enabled) {
-    // invoke a JNI call from the SIGQUIT handler
     JNIEnv *env;
-    int result = (*bsg_jvm)->GetEnv(bsg_jvm, (void **)&env, JNI_VERSION_1_4);
+    int result = (*bsg_jvm)->GetEnv(bsg_jvm, (void **) &env, JNI_VERSION_1_4);
 
     if (result == JNI_OK) { // already attached
-      bsg_notify_anr_detected(env);
+      (*env)->CallVoidMethod(env, obj_plugin, mthd_notify_anr_detected);
     } else if (result == JNI_EDETACHED) { // attach before calling JNI
       if ((*bsg_jvm)->AttachCurrentThread(bsg_jvm, &env, NULL) == 0) {
-        bsg_notify_anr_detected(env);
+        (*env)->CallVoidMethod(env, obj_plugin, mthd_notify_anr_detected);
         (*bsg_jvm)->DetachCurrentThread(
-            bsg_jvm); // detach to restore initial condition
+                bsg_jvm); // detach to restore initial condition
       }
     } // All other results are error codes
   }
+}
 
-  // Invoke previous handler, if a custom handler has been set
-  // This can be conditionally switched off on certain platforms (e.g. Unity)
-  // where this behaviour is not desirable due to Unity's implementation failing
-  // with a SIGSEGV
-  if (invokePrevHandler) {
-    struct sigaction previous = bsg_sigquit_sigaction_previous;
-    if (previous.sa_flags & SA_SIGINFO) {
-      previous.sa_sigaction(SIGQUIT, info, user_context);
-    } else if (previous.sa_handler == SIG_DFL) {
-      // Do nothing, the default action is nothing
-    } else if (previous.sa_handler != SIG_IGN) {
-      void (*previous_handler)(int) = previous.sa_handler;
-      previous_handler(signum);
-    }
+static void* sigquit_thread_main(void* _) {
+  while(!should_report_anr) {
+    usleep(100000);
   }
+  // Force at least one task switch after being triggered, ensuring that the signal masks are
+  // properly settled before triggering the Google handler.
+  usleep(10000);
+
+  notify_anr_detected();
+
+  bsg_google_anr_call();
+
+  // Give a little time for the Google handler to dump state, then exit this thread.
+  sleep(2);
+  return NULL;
 }
 
-/**
- * Configure ANR detection using a signal handler listening for SIGQUIT
- */
-void *bsg_monitor_anrs(void *_arg) {
+static void handle_sigquit(int signum, siginfo_t* info, void* user_context) {
+  // TODO: Save stack trace
```

Comment:
Does this still need to be addressed?

Reply:
Thats where the direct native-to-native call to stack unwind that I've been working on in the other integration PR will go. Right now it's not tracking native traces since that's not in the next branch.

Reply:
In the case the TODO should be removed so as not to indicate a partially complete job in the this release.

Is it just this TODO-comment that relates to the native ANR work or is there functionality that should be parred down for this release?

Reply:
It's the place in the code where the call to get a native stack trace will eventually go. I've updated the TODO to be more explicit about that, but I could remove it altogether if it's still subject to confusion...

Reply:
I don't see that it serves any purpose, as you don't need it to remind you long term, and makes the release look incomplete. I think it should be removed.